dsymbol/src/dsymbol/ufcs.d

```diff
@@ -26,43 +42,106 @@ enum string[string] INTEGER_PROMOTIONS = [
- *  $(LI is callable with $(D beforeDotSymbol) as it's first argument)
- * )
- *
- * Params:
- *     completionScope = current scope
- *     beforeDotSymbol = the symbol before the dot (implicit first argument to UFCS function)
- *     cursorPosition = current position
- * Returns:
- *     callable an array of symbols suitable for UFCS at $(D cursorPosition)
- */
-DSymbol*[] getSymbolsForUFCS(Scope* completionScope, const(DSymbol)* beforeDotSymbol, size_t cursorPosition)
+
+private TokenCursorResult getCursorToken(const(Token)[] tokens, size_t cursorPosition)
 {
-    if (beforeDotSymbol.isInvalidForUFCSCompletion)
+    auto sortedTokens = assumeSorted(tokens);
+    auto sortedBeforeTokens = sortedTokens.lowerBound(cursorPosition);
+
+    TokenCursorResult tokenCursorResult;
+
+    if (sortedBeforeTokens.empty) {
+        return tokenCursorResult;
+    }
+
+    if (sortedBeforeTokens.length >= 2 
+        && sortedBeforeTokens[$ - 1].type is tok!"."
+        && sortedBeforeTokens[$ - 2].type is tok!"identifier")
     {
-        return null;
+        // Check if it's UFCS dot completion
+        tokenCursorResult.completionContext = UFCSCompletionContext.DotCompletion;
+        tokenCursorResult.symbolIdentifierName = istring(sortedBeforeTokens[$ - 2].text);
+        return tokenCursorResult;
+    }
+    else
+    {
+        // Check if it's UFCS paren completion
+        size_t index = goBackToOpenParen(sortedBeforeTokens);
+
+        if (index == size_t.max)
+        {
+            return tokenCursorResult;
+        }
+
+        auto slicedAtParen = sortedBeforeTokens[0 .. index];
+        if (slicedAtParen.length >= 4
+            && slicedAtParen[$ - 4].type is tok!"identifier"
+            && slicedAtParen[$ - 3].type is tok!"."
+            && slicedAtParen[$ - 2].type is tok!"identifier"
+            && slicedAtParen[$ - 1].type is tok!"(")
+        {
+            tokenCursorResult.completionContext = UFCSCompletionContext.ParenCompletion;
+            tokenCursorResult.symbolIdentifierName = istring(slicedAtParen[$ - 4].text);
+            tokenCursorResult.functionName = istring(slicedAtParen[$ - 2].text);
+            return tokenCursorResult;
+        }
+
     }
+    // if none then it's unknown
+    return tokenCursorResult;
+}
+
+private void getUFCSSymbols(T, Y)(ref T localAppender, ref Y globalAppender, Scope* completionScope, size_t cursorPosition)
+
+{
 
     Scope* currentScope = completionScope.getScopeByCursor(cursorPosition);
-    assert(currentScope);
-    HashSet!size_t visited;
+    if (currentScope is null)
+    {
+        return;
+    }
 
-    // local appender
-    FilteredAppender!(a => a.isCallableWithArg(beforeDotSymbol), DSymbol*[]) localAppender;
+    DSymbol*[] cursorSymbols = currentScope.getSymbolsInCursorScope(cursorPosition);
+    if (cursorSymbols.empty)
+    {
+        return;
+    }
+
+    auto filteredSymbols = cursorSymbols.filter!(s => s.kind == CompletionKind.functionName).array;
```

Comment:
UFCS also works with aliased / imported names and templates (templated functions) - I saw they are resolved elsewhere, but are they included here? Might be worth a comment in code to mention what is being done here.

Reply:
This something I thought was needed, I will remove this piece of logic.

Reply:
no the logic might be fine here, I also didn't check if anything needs changing here. I just wanted a comment here to clarify why this only filters for function symbols (and e.g. not alias symbols, which are probably already resolved here?)

Reply:
the logic made the tests/tc_erroneous_body_content/expected1.txt show: `bar F`
which I at the time thought was right behavior, so I ended up scraping it, but if this is needed I will consider adding it in again :) but yeah you are right about alias symbols.